### PR TITLE
Issue 3396 - Single Sign On Login

### LINKED ISSUE #3396: Single Sign On: Login
### Description
Add the ability to log in via single sign on through AAD

This issue depends on https://github.com/3drepo/3drepo.io/issues/3395

### Goals
- [x] User should be able to log in with a Microsoft account if their 3drepo is linked with one
- [x] User should not be able to login with username/password if their account is setup for SSO
- [x] If a user tries to log in via SSO with an email that is associated with an account not linked up with SSO, they should be given an apprioriate error
- [ ] Forgot password should not send an email to a SSO linked user 
- [x] Server should respond with apprioriate error if AAD is not configured

#### To login via SSO:
route: `GET /sso/aad/authenticate`
This route should redirect the user to Microsoft to authenticate, which then redirects the user to the redirect endpoint

#### Redirect endpoint
route: `GET /sso/aad/authenticate-post`
This route should determine whether the account is associated with a 3drepo account (by means of getting the id and see if there's a matching user)

- If there is a match, we should login as per typical user login (i.e. record the login record, establish the session etc) and redirect the user back to whereever the request specified.
- If the email matched but there is no SSO linkage on the account, apprioriate error should return
- If there is no match at all, it should redirect to sign up as per https://github.com/3drepo/3drepo.io/issues/3395

#### Alteration to existing login
- If an account is associated with an SSO account, normal login route should not work for that user

#### Forgot password
- If a user try to trigger the forgot password journey with an account associated with SSO account, no email should be sent

### Tasks
To be filled in by the developer

### PR DESCRIPTION
**Pull request by ChristosTsiotsias**
This fixes #3396 

#### Description
Endpoint for Signle Sign On Login has been added
Unit and E2E tests added
Adjustment to SSO Signup to not send verification email to new users
Adjustment to Forgot Password to not send email to SSO users

#### Test cases
The user should be able to login with SSO provided they have signup up via SSO

### CODE REVIEW

File: backend/src/v5/middleware/dataConverter/inputs/sso/index.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+const { CryptoProvider } = require('@azure/msal-node');
+
+const Sso = {};
+
+Sso.addPkceProtection = async (req, res, next) => {
```
Comment (carmenfan) [inline backend/src/v5/middleware/dataConverter/inputs/sso/index.js:22]: this already exists in backend/src/v5/middleware/sso/pkce.js ?

File: backend/src/v5/middleware/dataConverter/inputs/users.js [hunk 2]
```diff
@@ -134,7 +139,12 @@ Users.validateForgotPasswordData = async (req, res, next) => {
 
 		try {
 			const usernameOrEmail = req.body.user;
-			const { user } = await getUserByUsernameOrEmail(usernameOrEmail, { user: 1, _id: 0 });
+			const { user, customData: { sso } } = await getUserByUsernameOrEmail(usernameOrEmail, { user: 1, _id: 0, 'customData.sso': 1 });
+
+			if (sso) {
```
Comment (carmenfan) [inline backend/src/v5/middleware/dataConverter/inputs/users.js:144]: I don't think we should be throwing error?

It should return as usual, just don't send email -this is(or at least, should be) the behaviour when someone is trying to trigger forgot password with an invalid email address/username

File: backend/src/v5/middleware/sso/aad.js [hunk 3]
```diff
@@ -14,19 +14,25 @@
...
 const { respond } = require('../../utils/responder');
-const { signupRedirectUri } = require('../../services/sso/aad/aad.constants');
 const { validateMany } = require('../common');
 
 const Aad = {};
 
+const redirectWithError = async (res, url, errorCode) => {
+	const urlRedirect = new URL(url);
+	urlRedirect.searchParams.set('error', errorCode);
+	await res.redirect(urlRedirect.href);
```
Comment (carmenfan) [inline backend/src/v5/middleware/sso/aad.js:33]: I don't think `res.redirect` returns a promise?

https://expressjs.com/en/api.html#res.redirect

File: backend/src/v5/middleware/sso/aad.js [hunk 4]
```diff
@@ -92,4 +96,30 @@ Aad.redirectToStateURL = (req, res) => {
 };
 Aad.authenticate = (redirectUri) => validateMany([addPkceProtection, authenticate(redirectUri)]);
 
+Aad.checkIfMsAccountIsLinkedTo3DRepo = async (req, res, next) => {
```
Comment (carmenfan) [inline backend/src/v5/middleware/sso/aad.js:99]: This is very wordy

something like `hasAssociatedAccount` should suffice?

File: backend/src/v5/middleware/sso/aad.js [hunk 5]
```diff
@@ -92,4 +96,30 @@ Aad.redirectToStateURL = (req, res) => {
...
+				req.body.user = user;
+				req.loginData = await recordSuccessfulAuthAttempt(user);
+				await next();
+			}
+		} catch {
+			redirectWithError(res, state.redirectUri, errorCodes.userNotFound);
+		}
+	} catch (err) {
+		logger.logError(`Failed to fetch user details from Microsoft API: ${err.message}`);
+		respond(req, res, templates.unknown);
```
Comment (carmenfan) [inline backend/src/v5/middleware/sso/aad.js:121]: this should also be a redirect?

File: backend/src/v5/middleware/sso/aad.js [hunk 6]
```diff
@@ -92,4 +96,30 @@ Aad.redirectToStateURL = (req, res) => {
...
 
+Aad.checkIfMsAccountIsLinkedTo3DRepo = async (req, res, next) => {
+	try {
+		const state = JSON.parse(req.query.state);
+
+		const { data: { id, mail } } = await getUserDetails(req.query.code, authenticateRedirectUri,
+			req.session.pkceCodes?.verifier);
+
+		try {
+			const { user, customData: { sso } } = await getUserByEmail(mail, { _id: 0, user: 1, 'customData.sso': 1 });
```
Comment (carmenfan) [inline backend/src/v5/middleware/sso/aad.js:107]: refactor the check into models/users

This is too into the guts for a middleware to do. Just extract the data and call something like `GetUserBySSOInfo(email, ssoId)` or something

File: backend/src/v5/models/users.js [hunk 7]
```diff
@@ -31,7 +31,7 @@ const COLL_NAME = 'system.users';
 const userQuery = (query, projection, sort) => db.findOne(USERS_DB_NAME, COLL_NAME, query, projection, sort);
 const updateUser = (username, action) => db.updateOne(USERS_DB_NAME, COLL_NAME, { user: username }, action);
 
-const recordSuccessfulAuthAttempt = async (user) => {
+User.recordSuccessfulAuthAttempt = async (user) => {
```
Comment (carmenfan) [inline backend/src/v5/models/users.js:34]: theses are all gone as of #3694

File: backend/src/v5/models/users.js [hunk 8]
```diff
@@ -227,14 +226,14 @@ User.addUser = async (newUserData) => {
 		},
 		permissions: [],
 		...deleteIfUndefined({ sso: newUserData.sso }),
```
Comment (carmenfan) [inline backend/src/v5/models/users.js:228]: I don't think you need `deleteIfFalse` or `deleteIfUndefined`
```suggestion
		...(newUserData.sso ? { sso: newUserData.sso } : { inactive: true } ),
```

File: backend/tests/v5/e2e/routes/sso/aad/aad.test.js [hunk 9]
```diff
@@ -71,8 +76,47 @@ const testAuthenticate = () => {
 
 const testAuthenticatePost = () => {
 	describe('Authenticate Post', () => {
+		test('should fail if state is not provided', async () => {
+			const res = await agent.get('/v5/sso/aad/authenticate-post')
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/sso/aad/aad.test.js:80]: should redirect?

File: backend/tests/v5/e2e/routes/sso/aad/aad.test.js [hunk 10]
```diff
@@ -71,8 +76,47 @@ const testAuthenticate = () => {
...
+				.expect(templates.unknown.status);
+			expect(res.body.code).toEqual(templates.unknown.code);
+		});
+
+		test('should redirect and add error to the query if user does not exist', async () => {
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/sso/aad/aad.test.js:85]:
```suggestion
		test('should redirect with ${errorCodes.userNotFound} if user does not exist', async () => {
```

File: backend/tests/v5/e2e/routes/sso/aad/aad.test.js [hunk 11]
```diff
@@ -71,8 +76,47 @@ const testAuthenticate = () => {
...
+			const userDataFromAad = { mail: generateRandomString(), id: generateRandomString() };
+			const state = { redirectUri: generateRandomURL() };
+			Aad.getUserDetails.mockResolvedValueOnce({ data: userDataFromAad });
+			const res = await agent.get(`/v5/sso/aad/authenticate-post?state=${encodeURIComponent(JSON.stringify(state))}`)
+				.expect(302);
+			expect(res.headers.location).toEqual(`${state.redirectUri}?error=${errorCodes.userNotFound}`);
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/sso/aad/aad.test.js:91]: Can we add a test to ensure if there's query string on the provided redirectUri we're retaining it?

i..e if the redirectUri is `https://www.abc.com/?blah=blah` then these error codes should still append correctly

File: backend/tests/v5/e2e/routes/sso/aad/aad.test.js [hunk 12]
```diff
@@ -71,8 +76,47 @@ const testAuthenticate = () => {
...
+		});
+
+		test('should redirect and add error to the query if user is a non SSO user', async () => {
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/sso/aad/aad.test.js:94]: again be specific about which error we're expecting on the test name

File: backend/tests/v5/e2e/routes/sso/aad/aad.test.js [hunk 13]
```diff
@@ -170,20 +214,20 @@ const signupPost = () => {
...
+		test('should sign a new user up and set them to active', async () => {
 			const state = { ...newUserData };
 			Aad.getUserDetails.mockResolvedValueOnce({ data: newUserDataFromAad });
 			await agent.get(`/v5/sso/aad/signup-post?state=${encodeURIComponent(JSON.stringify(state))}`)
 				.expect(302);
 
-			// ensure user is created
-			await getUserByUsername(newUserData.username, { _id: 1 });
+			// ensure user is created and is active
+			const user = await getUserByUsername(newUserData.username, { _id: 1, customData: 1 });
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/sso/aad/aad.test.js:229]:
```suggestion
			const user = await getUserByUsername(newUserData.username, { _id: 1, "customData.inactive": 1 });
```

File: backend/tests/v5/e2e/routes/users.test.js [hunk 14]
```diff
@@ -524,6 +530,11 @@ const testForgotPassword = () => {
 			expect(res.body.code).toEqual(templates.invalidArguments.code);
 		});
 
+		test('should not send email but return ok if user is an SSO user', async () => {
```
Comment (carmenfan) [inline backend/tests/v5/e2e/routes/users.test.js:533]: if you're really doing what this test says is doing, you should be mocking the mailer to make sure it's not called 😆 

As it stands it only tests if it returns ok if it's a SSO user

File: backend/tests/v5/unit/middleware/dataConverter/inputs/sso/sso.test.js [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+const { src } = require('../../../../../helper/path');
+
+const Sso = require(`${src}/middleware/dataConverter/inputs/sso`);
+
+const testAddPkceProtection = () => {
```
Comment (carmenfan) [inline backend/tests/v5/unit/middleware/dataConverter/inputs/sso/sso.test.js:22]: again i think this page already exist elsewhere

File: backend/tests/v5/unit/middleware/dataConverter/inputs/users.test.js [hunk 16]
```diff
@@ -64,7 +65,9 @@ UsersModel.getUserByQuery.mockImplementation((query) => {
 
 UsersModel.getUserByUsernameOrEmail.mockImplementation((usernameOrEmail) => {
```
Comment (carmenfan) [inline backend/tests/v5/unit/middleware/dataConverter/inputs/users.test.js:66]: Can we refactor this to do mockImplementationOnce/mockResolvedValueOnce within the test?

File: backend/tests/v5/unit/processors/users.test.js [hunk 17]
```diff
@@ -235,9 +235,12 @@ const testSignUp = () => {
 			email: generateRandomString(),
 			password: generateRandomString(),
 			firstName: generateRandomString(),
+			lastName: generateRandomString(),
+			company: generateRandomString(),
+			mailListOptOut: true,
 		};
 
-		test('should sign a user up', async () => {
+		test('should sign a user up and send verification email (non SSO user)', async () => {
```
Comment (carmenfan) [inline backend/tests/v5/unit/processors/users.test.js:243]: should add a line here to make sure USER_VERIFIED isn't triggered

File: backend/src/v5/middleware/dataConverter/inputs/users.js (hunk 2 above)
Comment (ChristosTsiotsias) [reply backend/src/v5/middleware/dataConverter/inputs/users.js:144]: Its being caught in line 151 and returns OK

**Review by carmenfan — CHANGES_REQUESTED** [review]
4 bugs:

#### 1. PKCE is only appended to the session if the user call authenticate first
If the user wishes the sign up, they have to call authenticate endpoint first or the PKCE will not establish, and acquire token will fail. Whilst this is likely to be the steps, we shouldn't be assuming it. The sign up route should also try to append PKCE if it doesn't exist.

#### 2. SSO users are able to be logged into multiple sessions
Currently, if you log into 3drepo via SSO on a browser, then log in again on a different browser (or incognito mode), both sessions stays valid. 

Like logging in with username and password, the first session should get kicked out as soon as the user logged in elsewhere.

#### 3. LoginRecords are not showing the username 
If user logged in via SSO it's showing up as null

#### 4. We're not logging referrer if the user log in via SSO
Is it possible to retain the referrer (pick it out from the request and retain it in the state?) - this might be the cause for 2 actually.
image



image

**Comment by ChristosTsiotsias** [comment]
These are the settings we need for the AAD app to make sure the auth token includes the user info we need

image

**Review by carmenfan — CHANGES_REQUESTED** [review]
@ChristosTsiotsias still missing referrer:
image

**Comment by ChristosTsiotsias** [comment]
> @ChristosTsiotsias still missing referrer: image

it does have a referrer when I test it :/

I test it with this url on a browser https://3drepo.local/api/v5/sso/aad/authenticate?redirectUri=https://3drepo.local/v5/dashboard 



> @ChristosTsiotsias still missing referrer: image

It seems to have the referrer when I test it 

image

I test it by running https://3drepo.local/api/v5/sso/aad/authenticate?redirectUri=https://3drepo.local/v5/dashboard on the browser and following the steps

**Comment by carmenfan** [comment]
> > @ChristosTsiotsias still missing referrer: image
> 
> it does have a referrer when I test it :/
> 
> I test it with this url on a browser https://3drepo.local/api/v5/sso/aad/authenticate?redirectUri=https://3drepo.local/v5/dashboard
> 
> > @ChristosTsiotsias still missing referrer: image
> 
> It seems to have the referrer when I test it
> 
> image
> 
> I test it by running https://3drepo.local/api/v5/sso/aad/authenticate?redirectUri=https://3drepo.local/v5/dashboard on the browser and following the steps

even if it works for you, login.live.com is not the desired referer.... 😆

**Review by carmenfan — APPROVED** [review]

**Pull request merged**